src/crimson/osd/pg_backend.cc

```diff
@@ -227,23 +250,25 @@ seastar::future<bufferlist> PGBackend::read(const object_info_t& oi,
     // read size was trimmed to zero and it is expected to do nothing,
     return seastar::make_ready_future<bufferlist>();
   }
-  std::optional<uint32_t> maybe_crc;
-  if (oi.is_data_digest() && offset == 0 && length >= oi.size) {
-    maybe_crc = oi.data_digest;
-  }
-  return _read(oi.soid, offset, length, flags).then(
-    [maybe_crc, soid=oi.soid, size=oi.size](auto bl) {
-      // whole object?  can we verify the checksum?
-      if (maybe_crc && bl.length() == size) {
-        if (auto crc = bl.crc32c(-1); crc != *maybe_crc) {
-          logger().error("full-object read crc {} != expected {} on {}",
-            crc, *maybe_crc, soid);
-          // todo: mark soid missing, perform recovery, and retry
-          throw ceph::osd::object_corrupted{};
-        }
-      }
-      return seastar::make_ready_future<bufferlist>(std::move(bl));
-    });
+  return _read(oi.soid, offset, length, flags).safe_then(
+    [&oi](auto bl) {
+      return _read_verify_data(oi, bl).safe_then(
+        [bl = std::move(bl)] {
+          return seastar::make_ready_future<bufferlist>(std::move(bl));
+        }, verify_data_errorator::pass_further{});
+    }, ll_read_errorator::pass_further{});
```

Comment:
Hmm, maybe error handlers (like `pass_further`) shouldn't be part of `errorator` at all. Instead programmer would need to specify errors manually: `ceph::pass_further<error1, error2>{}`. If so, adding a new error into e.g. `ll_read_errorator` would require revisiting all those bulky handlers.

Reply:
Can we arrange it so that safe_then with one argument silently passes the same errors through?

Reply:
Seems possible.

Reply:
yeah, i concur with Sam.